Switch LWRPs to custom resources

damacus opened the pull request:
### Description

This PR will replace all current LWRPs with custom resources. 
And as part of the "unified interface" philosophy I've been taking recently, node attributes will be removed. 

It should make the cookbook easier to write, amend & consume going forward. 



# Tests
Again it should make it easier to test everything: 

- [x] Test suites for all resources
  - [x] gem
  - [x] global
  - [x] plugin
  - [x] rehash
  - [x] ruby
  - [x] script
  - [x] user_install
  - [x] system_install




### Issues Resolved

#170 

### Contribution Check List

- [x] All tests pass.
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable

### Fixed
- [x] Test for Nokogiri gem usage
- [x] Test for OpenSSL usage

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sous-chefs/ruby_rbenv/169)
<!-- Reviewable:end -->

damacus committed:
Switch global and plugin to custom resource
Add ignore for foodcritic 085
Update some bats tests to inspec

damacus committed:
Convert more inspec tests for global Ruby

damacus committed:
Add integration tests for installation.
Move inspec controls to their own directory and setup an inspec profile

damacus committed:
Move inspec tests into the correct dir

tas50 commented on Gemfile:
```diff
@@ -6,6 +6,5 @@
 
-gem 'tomlrb'
 gem 'stove'
 gem 'community_cookbook_releaser'
```
This should go away as well

damacus committed:
Source rbenv script before running command now works in conjunction with the Bash
resource.
Remove toml generator

damacus committed:
Fix rubocop

damacus committed:
Adds AppVeyor testing.

Relates to #171

damacus committed:
Just use the IIS cookbook's appveyor.yml

damacus committed:
Fix AmbiguousRegexpLiteral

damacus committed:
Remove deleted providers

damacus committed:
Add script, ruby and rehash resrouces

damacus committed:
Fix patch command

damacus committed:
Be explicit with provides because the resource name is different to the
cookbook name so we can't rely on that.

Use BeginRescueEnd for the warning, so we don't reinstall Ruby several
times

damacus committed:
Remove unit tests that *only* tested whether we defaulted
attributes/properties or not

damacus committed:
Stop using embedded path

damacus committed:
Removing a bunch of stuff from the README.

This needs to be started from scratch with the new resources.

Some examples added to tests would work as a starter

damacus committed:
Merge branch 'master' into update_tests

damacus committed:
Fix unit tests

damacus committed:
Tests now all passing

tas50 commented on resources/global.rb:
```diff
@@ -19,15 +19,31 @@
+  include Chef::Rbenv::ScriptHelpers
 
-def to_s
-  "#{super} (#{@user || 'system'})"
```
what was the intention here?

damacus replied:
Honestly. I don't know. I think it's a hangover from the lwrp. Fletcher might know more *possibly*

tas50 commented on resources/plugin.rb:
```diff
@@ -1,15 +1,32 @@
+
+provides :rbenv_plugin
+
+property :name, String, name_property: true
```
This line can go away

tas50 commented on resources/rehash.rb:
```diff
@@ -19,11 +19,23 @@
 
-actions :run
-default_action :run
+property :name, String, name_property: true
```
This one can go away as well. I'll have a foodcritic rule for it some day

tas50 commented on resources/ruby.rb:
```diff
@@ -19,25 +20,88 @@
+    Chef::Log.debug("#{new_resource} build time was " \
+      "#{(Time.now - install_start) / 60.0} minutes")
+  rescue
+    Chef::Log.warn('Ruby version already installed')
```
Should this be a warning?

damacus replied:
Hmmm good point. I'll change it to info

tas50 commented on resources/script.rb:
```diff
@@ -19,21 +19,70 @@
 
-actions :run
-default_action :run
+property :name, String, name_property: true
```
This can go away

tas50 commented on test/integration/system_ruby/controls/global_ruby.rb:
```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+global_ruby = '2.1.6'
```
We should probably get this onto a support Ruby release

tas50 reviewed (dismissed):
There's a few things to cleanup here. Huge improvement though

damacus committed:
Update global ruby version to 2.3.4 (current oldest supported version)

Start updating the gem resource.
remove a bunch of name, name properties

damacus committed:
Fix action_class

damacus committed:
Add notification for travis for build

sekrett commented on metadata.rb:
```diff
@@ -14,7 +14,6 @@
 supports 'ubuntu'
 supports 'linuxmint'
 supports 'debian'
-supports 'freebsd'
```
Remember to remove it from readme too.

damacus replied:
Added it back in ☺️ pretty sure we can test it

damacus committed:
Update driver config, fix logic on ruby install

damacus committed:
Add gem_install tests.

damacus committed:
Add rbnev_system_install resource

damacus committed:
Update CHANGELOG

damacus committed:
Add package dep library. So we don't ahve to rely on the node attributes
on an external cookbook.
Add more actions to the gem install.
Tidy up the global install. This will break/change existing recipe based
installs

tas50 commented on .kitchen.yml:
```diff
@@ -3,7 +3,7 @@ driver:
 
 provisioner:
   name: chef_zero
-  deprecations_as_errors: true
```
We still want this for the next round of deprecations

damacus replied:
fixed

tas50 commented on README.md:
```diff
@@ -1,30 +1,29 @@
-# ruby_rbenv Chef Cookbook
```
This is the actual name until we can get the rbenv namespace

tas50 commented on README.md:
```diff
@@ -1,30 +1,29 @@
-The following platforms have been tested with this cookbook, meaning that the recipes and LWRPs run on these platforms without error:
+### Platform
 
 - Arch
```
I’d take out arch since we can’t test it

tas50 commented on README.md:
```diff
@@ -1,30 +1,29 @@
-- Freebsd
+- Debian derivatives (Debian 7+, Ubuntu 12.04+, Linux Mint)
 - Fedora
 - Gentoo
```
Same with gentoo

tas50 commented on README.md:
```diff
@@ -1,30 +1,29 @@
+- Debian derivatives (Debian 7+, Ubuntu 12.04+, Linux Mint)
 - Fedora
 - Gentoo
 - Mac OS X
```
This is macOS now

tas50 commented on README.md:
`````diff
@@ -148,31 +136,32 @@ Determines how to handle installing updates to the rbenv. There are currently 2
 node.default['rbenv']['rubies'] = [ "1.9.3-p0", "jruby-1.6.5" ]
-```
+`
+````
`````
Typo?

damacus replied:
The README in general will need a total fresh start. I imagine the resources will be completely different.

tas50 commented on README.md:
```diff
@@ -185,9 +174,10 @@ node.default['rbenv']['rubies'] = [ "1.9.3-p0", "jruby-1.6.5",
   :environment => { 'CFLAGS' => '-march=native -O2 -pipe' }
   }
 ]
+`
```
?

tas50 commented on libraries/package_deps.rb:
```diff
@@ -0,0 +1,21 @@
+def jruby_package_deps
```
We need to keep these out of the global namespace. Someone will choose the same name and then they’ll collide

damacus replied:
Good point!

tas50 commented on libraries/package_deps.rb:
```diff
@@ -0,0 +1,21 @@
+  when 'debian'
+    %w(make g++)
+  when 'freebsd'
+    %w( alsa-lib bash dejavu expat fixesproto fontconfig freetype2 gettext-runtime giflib indexinfo inputproto java-zoneinfo javavmwrapper kbproto libICE libSM libX11 libXau libXdmcp libXext libXfixes libXi libXrender libXt libXtst libfontenc libpthread-stubs libxcb libxml2 mkfontdir mkfontscale openjdk8 recordproto renderproto xextproto xproto )
```
How did we end up with X11 here?

damacus replied:
hell knows. Honestly I'd love to go through this package list at some point and slim it down. 

I'll leave this in TODO: for now.

tas50 commented on metadata.rb:
```diff
@@ -14,7 +14,6 @@
 supports 'ubuntu'
 supports 'linuxmint'
 supports 'debian'
-supports 'freebsd'
```
?

damacus replied:
nope I was imagining that for some reason. I'll remove it, and note it in the readme, as there's no way we can test it reliably.

damacus committed:
Add user_install resource

damacus committed:
gah omg the bash hax

tas50 reviewed (dismissed):
I added a few notes

damacus committed:
Remove unsupportable/untestable platforms
Add the git dir as the actual user

damacus committed:
owner --> user

damacus committed:
Move package deps into the owner namespace

damacus committed:
We don't need user home anymore ( ~user instead)

damacus committed:
Yes we do... but we shouldn't ne using set_updated

damacus committed:
Remove recipes. Add a changelog entry for usage and where all the things have gone!

damacus committed:
* Tidies up rspec
* Add a few kitchen suites
* global root_path is now calculated because on either user or system install
*

damacus committed:
root_path helper that's populated by either user or global resources.

* deprecates root_path ina lot of places
* Update package deps according to https://github.com/rbenv/ruby-build/wiki

damacus committed:
cookstyle

damacus committed:
Remove foodcritic exception

damacus committed:
Update CHANGELOG

damacus committed:
Update README!
Get rid of a bunch of properties that are not needed

damacus committed:
Remove the Ruby gems package resource.

Fingers crossed we don't need it. I'm not 100% sure this will actually
work... More tests required

damacus committed:
Remove a bunch of suites that we don't actually need (they're tested as part of other suites)

update dokken config.

damacus committed:
Update travis config

damacus committed:
Update dokken testing (new debian!)

Add curl to some dists

damacus committed:
Fix macOS in readme

damacus committed:
squash - CHANGELOG

damacus committed:
Fix user install resource for dokken

damacus committed:
Add dokken recipe to make dokken work with user installs

damacus committed:
Fix opensuse package deps

damacus committed:
Add known bugs section stating fedora and 2.3.1

damacus committed:
Add more dokken fixes

damacus committed:
Switch to dokken images

damacus committed:
Tidy up dokken config add make to package deps on centos

damacus committed:
sudo for debian-8

damacus committed:
Add sudo to a bunch more distros. Except 14.04

damacus committed:
Fix more dokken config

damacus committed:
Update gem tests.

damacus committed:
Live_stream the bash resource

damacus committed:
Default live stream to true for debug.
stop xcluding centos :'(

damacus committed:
Exclude centos-6 due to long compile time.
Add sudo for opensuse

damacus committed:
Package deps for opensuse resolved
install both curl and tar on dokken/opensue
Add verbose flag for streamed debugged output.

damacus committed:
Re-add centos to doken

damacus committed:
Update tetsts for global

damacus committed:
cookstyle

damacus committed:
Fix deprection warning in test

damacus committed:
Verboser?

damacus committed:
Remove centos 6 platform on travis. Build slave gets terminated...

tas50 commented on .kitchen.dokken.yml:
```diff
@@ -14,86 +14,74 @@ verifier:
     pid_one_command: /bin/systemd
     intermediate_instructions:
-      - RUN zypper --non-interactive install aaa_base perl-Getopt-Long-Descriptive
+    - RUN zypper --non-interactive install sudo curl tar
```
I merged the PR to dokken-images. This can get removed now

tas50 commented on .travis.yml:
```diff
@@ -19,22 +18,47 @@ services: docker
-  - /opt/chefdk/embedded/bin/chef --version
-  - /opt/chefdk/embedded/bin/cookstyle --version
-  - /opt/chefdk/embedded/bin/foodcritic --version
+  - /opt/chefdk/bin/chef --version
```
These don't actually need to be absolute paths. chef --version is fine

tas50 commented on .travis.yml:
```diff
@@ -19,22 +18,47 @@ services: docker
       env: UNIT_AND_LINT=1
+
+notifications:
+    slack: northops:FsfnzghB9NkVWXanv0RivJZZ
```
where is this going?

damacus replied:
my other team for notifications 👍

tas50 commented on README.md:
```diff
@@ -1,30 +1,27 @@
+### Platform
 
-### Chef
+- Debian derivatives (Debian 7+, Ubuntu 12.04+, Linux Mint)
```
We should remove the mention of Ubuntu 12.04. Also who knows if linux mint works. It probably does, but we're not testing it

tas50 commented on README.md:
```diff
@@ -1,30 +1,27 @@
+- macOS
+- RHEL derivatives (RHEL, CentOS, Amazon Linux, Oracle, Scientific Linux)
 
 This cookbook requires Chef 12.1+.
```
This is mentioned above as 12.9+

tas50 commented on README.md:
```diff
@@ -33,541 +30,112 @@ This cookbook requires Chef 12.1+.
 - [ruby_build cookbook][ruby_build_cb]
 - java cookbook if installing jruby (recommended, but not required in the metadata)
```
We don't actually recommend this in the metadata anymore

tas50 commented on README.md:
```diff
@@ -33,541 +30,112 @@ This cookbook requires Chef 12.1+.
 - [ruby_build cookbook][ruby_build_cb]
```
This is gone in the metadata

tas50 commented on README.md:
````diff
@@ -33,541 +30,112 @@ This cookbook requires Chef 12.1+.
 
 ```ruby
-node.default['rbenv']['rubies'] = [ "1.9.3-p0", "jruby-1.6.5" ]
+rbenv_gem 'gem_name' do
````
These samples seem kinda confusing and I think a lot of people are going to try to use the commas in them. Also most people have no idea what a name_property is

tas50 commented on README.md:
```diff
@@ -33,541 +30,112 @@ This cookbook requires Chef 12.1+.
+Runs a rbenv aware script.
 
-The default is an empty array: `[]`.
+``ruby rbenv_script 'foo' do rbenv_version, String code, String creates, String cwd, String environment, Hash group, String path, Array returns, Array, default: [0] timeout, Integer user, String umask, [String, Integer] end
```
This ruby block isn't rendering

tas50 commented on metadata.rb:
```diff
@@ -4,23 +4,18 @@
-
-depends 'ruby_build', '>= 1.1.0' # if using the rbenv LWRP, ruby-build must be installed
+version '2.0.0'
+chef_version '>= 12.19' if respond_to?(:chef_version)
```
The readme says 12.9+. Is this a typo?

tas50 commented on metadata.rb:
```diff
@@ -4,23 +4,18 @@
 supports 'fedora'
 supports 'amazon'
 supports 'scientific'
 supports 'suse'
```
If opensuse works that's opensuse and opensuseleap

tas50 commented on resources/system_install.rb:
```diff
@@ -0,0 +1,72 @@
+      %w(git)
+    when 'freebsd'
+      %w(git bash)
+    when 'gentoo'
```
This can get collapsed with mac_os_x

tas50 commented on resources/system_install.rb:
```diff
@@ -0,0 +1,72 @@
+      %w(git bash)
+    when 'gentoo'
+      %w(git)
+    when 'arch'
```
This can get collapsed with rhel

damacus committed:
Remove tar/curl as we've fixed this upstream

damacus committed:
Update travis paths

Update readme with more readable examples
Update platform deps

damacus committed:
Test with chef 12.9 instead of 12 latest

More readme fixes

damacus committed:
Don't require stove, as it's in the chefdk now

damacus committed:
Remove provides

damacus committed:
Re-add provides as we need this to map the resource name correctly(!)

damacus committed:
Found a matching 12.9 to test

damacus committed:
Cookstyle ftw

shoekstra reviewed (approved):
A lot to digest but the InSpec tests look reasonable and are passing in Travis so 👍  from me.
